src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/LambdaCompiler.Generated.cs

```diff
@@ -12,6 +14,24 @@ partial class LambdaCompiler
         [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Maintainability", "CA1505:AvoidUnmaintainableCode"), System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Maintainability", "CA1502:AvoidExcessiveComplexity")]
         private void EmitExpression(Expression node, CompilationFlags flags)
         {
+            // When compling deep trees, we run the risk of triggering a terminating StackOverflowException,
+            // so we probe for sufficient stack here. When we run out of stack, we simply continue the work
+            // on another thread. Note that EnsureSufficientExecutionStack is very fast, so the overhead of
+            // this overflow protection is minimal.
```

Comment:
I had the same concern about possibly recursive trees but couldn't (yet) come up with a way to construct them; the factory methods and `Update` methods don't lend themselves to this kind of evil, and the trees are immutable.

Maybe the right balance here is to be found by defining a range. The lower bound is where we start worrying about probing, and some (high enough) upper bound is where we start to panic because the user may have created a large enough tree (or created a cycle in some way) to cause insanely deep recursion. In the latter case, we could stop catching the `InsufficientExecutionStackException` and let it propagate (in lieu of `StackOverflowException`). The two values could be the lower bound (say 100) and the number of times we allow to jump to another thread to continue (say 10).

I'll have a look at what such an approach could look like.